core/domain/html_cleaner.py

```diff
@@ -782,11 +783,202 @@ def add_caption_attr_to_image(html_string):
+
+    rich_text_components = feconf.RICH_TEXT_COMPONENTS
+    # Mapping of type to classes defined in objects.py.
+    class_dict = {
+        'SanitizedUrl': objects.SanitizedUrl,
+        'UnicodeString': objects.UnicodeString,
+        'MathLatexString': objects.MathLatexString,
+        'Filepath': objects.Filepath,
+        'Int': objects.Int,
+        'Boolean': objects.Boolean,
+        'Html': objects.Html
+    }
+    filename_re = r'^[A-Za-z0-9+/]*\.((png)|(jpeg)|(gif)|(jpg))$'
+
+    for component in rich_text_components:
+        # This flag is set to true if the html string is invalid.
+        is_invalid = False
+        tag_name = component['name']
+        customization_args = component['customization_args']
+        required_attrs_names = [arg['name'] for arg in customization_args]
+
+        for tag in soup.findAll(name=tag_name):
+            attrs = tag.attrs
+            attrs_names = attrs.keys()
+            # Ensure that no tag attribute is missing and also check
+            # there is no extra attribute.
+            if set(attrs_names) != set(required_attrs_names):
+                is_invalid = True
+                break
+
+            for customization_arg in customization_args:
+                arg_name = customization_arg['name']
+                arg_value = json.loads(unescape_html(attrs[arg_name]))
+                if customization_arg['type'] != 'custom':
+                    arg_type = class_dict[customization_arg['type']]
+                    try:
+                        arg_type.normalize(arg_value)
+                        # To ensure filepath is in valid format.
+                        if arg_name == 'filepath-with-value':
+                            if not re.match(filename_re, arg_value):
+                                is_invalid = True
+
+                        # To ensure video id has 11 chars.
+                        elif arg_name == 'video_id-with-value':
+                            if len(arg_value) != 11:
+                                is_invalid = True
+
+                        # To ensure that nested collapsible or tab
+                        # is not present in collapsible component.
+                        elif arg_name == 'content-with-value':
+                            inner_soup = bs4.BeautifulSoup(
+                                arg_value.encode(encoding='utf-8'),
+                                'html.parser')
+                            collapsible_count = len(inner_soup.findAll(
+                                name='oppia-noninteractive-collapsible'))
+                            tabs_count = len(inner_soup.findAll(
+                                name='oppia-noninteractive-tabs'))
+                            if collapsible_count or tabs_count:
+                                is_invalid = True
+
+                    except Exception:
+                        is_invalid = True
+                else:
+                    # This block is to check args for tab component.
```

Comment:
Same thing -- there needs to be some reference to tabs somewhere in the "if" check (or to the type of the item).

Reply:
Added `tag_name` in if check (a complete else block for tabs is present and this check is within that block)